lib/fluent/command/unpacker.rb

```diff
@@ -0,0 +1,253 @@
+#
+# Fluentd
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'optparse'
+require 'msgpack'
+
+require 'fluent/msgpack_factory'
+require 'fluent/formatter'
+require 'fluent/plugin'
+require 'fluent/config/element'
+
+class FluentUnpacker
+  SUBCOMMAND = %w(cat head formats)
+  HELP_TEXT = <<HELP
+Usage: fluent-unpacker <command> [<args>]
+
+Commands of fluent-unpacker:
+   cat     :     Read files sequentially, writing them to standard output.
+   haed    :     Display the beginning of a text file.
+   foramts :     Display plugins that you can use.
+
+See 'fluent-unpacker <command> --help' for more information on a specific command.
+HELP
+
+  def initialize(argv = ARGV)
+    @argv = argv
+  end
+
+  def call
+    command_class = Command.const_get(command)
+    command_class.new(@argv).call
+  end
+
+  private
+
+  def command
+    command = @argv.shift
+    if !command || !SUBCOMMAND.include?(command)
+      usage "Required subcommand : #{SUBCOMMAND.join(' | ')}"
+    end
+
+    command.split('_').map(&:capitalize).join('')
+  end
+
+  def usage(msg = nil)
+    puts HELP_TEXT
+    puts "Error: #{msg}" if msg
+    exit 1
+  end
+end
+
+module Command
+  class Base
+    def initialize(argv = ARGV)
+      @argv = argv
+
+      @options = {}
+      @opt_parser = OptionParser.new do |opt|
+        opt.separator 'Options:'
+
+        opt.on('-p DIR', '--plugins', 'add library path') do |v|
+          @options[:plugins] = v
+        end
+      end
+    end
+
+    def call
+      raise NotImplementedError, 'BUG: command  MUST implement this method'
+    end
+
+    private
+
+    def usage(msg = nil)
+      puts @opt_parser.to_s
+      puts "Error: #{msg}" if msg
+      exit 1
+    end
+
+    def parse_options!
+      ret = @opt_parser.parse(@argv)
+
+      if @options[:plugins] && !Dir.exist?(@options[:plugins])
+        usage "Directory #{@options[:plugins]} doesn't exist"
+      elsif @options[:plugins]
+        Fluent::Plugin.add_plugin_dir(@options[:plugins])
+      end
+
+      ret
+    end
+  end
+
+  module Formattable
+    DEFAULT_OPTIONS = {
+      format: :out_file
+    }
+
+    def initialize(argv = ARGV)
+      super
+
+      @options.merge!(DEFAULT_OPTIONS)
+      @params = {}
+      if i = @argv.index('--')
+        @params = @argv[i+1..-1].reduce({}) do |acc, e|
+          k, v = e[1..-1].split('=')
```

Comment:
`e[1..-1]` is used for trimming prefix(`-`) in ` -fields=message`.
it is used like this.
 `$ fluent-unpacker cat --formats=csv  packed.log -- -fields=message`

But, I removed it .